### ISSUE TITLE
How to use custom Google Geocoding API key?

### ISSUE THREAD

**Issue by jplehmann** [furious-luke/django-address#37]
My data migration failed in production (on Heroku) with:

```
'The given key has gone over the requests limit in the 24'
geopy.exc.GeocoderQuotaExceeded: The given key has gone over the requests limit in the 24 hour period or has submitted too many requests in too short a period of time.
```

Presumably because I'm not really using a key but it's based on IP which is shared in the cloud.

I got around this by writing a managment command with the same code, which was not in a single transaction so it could convert a few at a time before failing.  On the 3rd try it managed to eat a few hundred and complete... but before that it was only doing about 10-20 at a time.

Bottom line, seems like we'd like to supply our own key.  But I'm not 100% sure this is how it works.

**Comment by furious-luke** [comment]
Hi John,
I've added the ability to set your own Google API key. You can do it by supplying a value for `GOOGLE_API_KEY` in your site's settings file. Of course, you'll need to go to the Google API console and get yourself a key for your site first.

**Comment by jplehmann** [comment]
Thanks Luke, that's great!  Since I worked around this I won't be doing it
right away, but it's possible that I will hit other limits within the cloud
and this would give me a way to fix it.

**Comment by Altaisoft** [comment]
For some unknown reason, this widget does not work for me without a key. But, I cannot find any trace of `GOOGLE_API_KEY` setting in the code. Shall I make up a pull request implementing this functionality?

**Comment by cesarmanzo** [comment]
I don't think so... It works for me without the key, so I guess it would be better to just put something on the instructions about it.

**Comment by Altaisoft** [comment]
In JavaScript console, Google libraries are complaining:

    Google Maps API error: MissingKeyMapError https://developers.google.com/maps/documentation/javascript/error-messages#missing-key-map-error
    Google Maps API warning: NoApiKeys https://developers.google.com/maps/documentation/javascript/error-messages#no-api-keys

I am overcoming this with a custom widget, very quick and dirty solution:

    class KeyedAddressWidget(AddressWidget):
        @property
        def media(self):
            media = super(KeyedAddressWidget, self).media
            media._js[0] += '&key=' + getattr(settings, 'GOOGLE_API_KEY', '')
            return media

Does anybody experience problems like that?

**Comment by cesarmanzo** [comment]
Yes, I get the API warning, but it's harmless. When you are in production, you should have your API key linked anyway. If you don't want to see the warning, just put your API key and that's it.

**Comment by Altaisoft** [comment]
The thing is that defining `GOOGLE_API_KEY` in `settings` has no effect. This key is not used by the application. You can search by `GOOGLE_API_KEY` in repository, and the single place where it can be found is this thread. Maybe I am missing something, though?

**Comment by cesarmanzo** [comment]
You know, maybe I should shut up since I haven't tested that setting yet, I'm not in production yet.

**Comment by mliudev** [comment]
@Altaisoft It looks like GOOGLE_API_KEY is defined in the dev branch, but after testing it doesn't appear to work as expected. I created a PR here: https://github.com/furious-luke/django-address/pull/51 to address the issue. See if that works for you.

**Comment by Altaisoft** [comment]
Thank you for the commit, I haven't tested it yet, but I suppose that, if there is no `GOOGLE_API_KEY` in settings it will crash with `AttributeError`. `getattr(settings, 'GOOGLE_API_KEY', None)` would work.

**Comment by mliudev** [comment]
Good catch, I've pushed a fix that uses getattr to ensure the setting is not 'None'.

**Comment by emmurd** [comment]
I got the same problem in dev. A workaround I found is to use `127.0.0.1:8000` instead of `localhost`. The google api will allow the communication. But this fix will be mandatory to prod I think.

**Comment by banagale** [comment]
I believe this was completed in pull/59 as referenced in pull/52
